File: tools/bench/bounded_queue_if_vs_while.cpp
// 조건변수 대기를 `if` 로 감쌌을 때 실제로 무엇이 깨지는가 (C++).
//
// bounded_queue_if_vs_while.py 와 같은 실험이다. 소비자가 셋(코어 4)이고
// 생산자가 `notify_all` 로 전원을 깨우면, 깨어난 소비자가 락을 다시 잡는 사이에
// 다른 소비자가 먼저 가져간다. `if` 로 감싼 쪽은 그 사실을 확인하지 않고
// 빈 큐를 만난다 — **통지 도둑질**이다.
//
// 본문의 ::: dual 은 "있음/없음" 만 찍는다(실행마다 흔들리는 값을 지면에 박지 않기
// 위해서다). 이 스크립트는 그 뒤에 있는 재현율을 센다.
//
// 측정 환경: CLAUDE.md §1-3 (Ubuntu 24.04 / g++ 13.3.0 -std=c++17 -O2 / 4 코어).
// 빌드·실행: g++ -std=c++17 -O2 -pthread tools/bench/bounded_queue_if_vs_while.cpp -o /tmp/iw && /tmp/iw

#include <chrono>
#include <condition_variable>
#include <cstdio>
#include <deque>
#include <mutex>
#include <thread>
#include <vector>
using namespace std;

static const int ROUNDS = 200, ITEMS = 200, NCONS = 3;   // 소비자 3 / 코어 4

struct Round { int bad, got; };

Round one_round(bool recheck, int gap_us) {
    deque<int> box;
    bool done = false;
    int bad = 0, got = 0;
    mutex mu;
    condition_variable cv;

    auto producer = [&] {
        for (int i = 0; i < ITEMS; i++) {
            if (gap_us) this_thread::sleep_for(chrono::microseconds(gap_us));
            { lock_guard<mutex> g(mu); box.push_back(i); }
            cv.notify_all();
        }
        { lock_guard<mutex> g(mu); done = true; }
        cv.notify_all();
    };

    auto consumer = [&] {
        for (;;) {
            unique_lock<mutex> lk(mu);
            if (recheck) {
                while (box.empty() && !done) cv.wait(lk);
            } else {
                if (box.empty() && !done) cv.wait(lk);
            }
            if (box.empty()) {
                if (done) return;
                bad += 1;            // 빈 큐를 만났다 — 통지를 도둑맞았다
                continue;
            }
            box.pop_back();
            got += 1;
        }
    };

    vector<thread> ts;
    for (int k = 0; k < NCONS; k++) ts.emplace_back(consumer);
    thread p(producer);
    p.join();
    for (auto& t : ts) t.join();
    return {bad, got};
}

void survey(bool recheck, int gap_us, const char* label) {
    int rounds_bad = 0;
    long long total_bad = 0;
    for (int r = 0; r < ROUNDS; r++) {
        Round x = one_round(recheck, gap_us);
        if (x.got != ITEMS) { printf("항목 수가 어긋났다: %d\n", x.got); return; }
        total_bad += x.bad;
        rounds_bad += (x.bad != 0);
    }
    printf("%-36s %3d/%d 회 실패   빈 큐 조우 총 %lld회\n",
           label, rounds_bad, ROUNDS, total_bad);
}

int main() {
    printf("소비자 %d / 코어 4 / 항목 %d개 / %d회 반복\n", NCONS, ITEMS, ROUNDS);
    survey(false, 200, "if   판, 생산 간격 0.2ms");
    survey(true, 200, "while 판, 생산 간격 0.2ms");
    survey(false, 0, "if   판, 생산 간격 0 (큐가 안 빈다)");
    return 0;
}

File: tools/bench/bounded_queue_if_vs_while.py
"""조건변수 대기를 `if` 로 감쌌을 때 실제로 무엇이 깨지는가 (Python).

교과서는 "가짜 깨움(spurious wakeup) 때문에 `while` 로 감싸라" 고 쓴다.
리눅스에서 진짜 가짜 깨움은 드물다. 그러나 소비자가 둘 이상이면
**통지 도둑질**이 흔하게 일어난다 — 깨어난 소비자가 락을 다시 잡기 전에
다른 소비자가 먼저 가져가서, 깬 쪽이 빈 큐를 만난다.

`if` 판과 `while` 판을 각각 200회 돌려 실패 횟수를 센다. 생산 간격을 0으로 두면
큐가 빌 틈이 없어 `if` 판도 통과한다 — 부하가 낮으면 경쟁이 숨는다는 것을
같은 스크립트 안에서 보이기 위해 두 조건을 다 잰다.

측정 환경: CLAUDE.md §1-3 (Ubuntu 24.04 / CPython 3.13.12 / 4 코어).
실행: python3.13 tools/bench/bounded_queue_if_vs_while.py
"""

import threading
import time

ROUNDS = 200
ITEMS = 200
CONSUMERS = 3      # 코어 4


def one_round(recheck, gap):
    box = []
    cv = threading.Condition()
    done = [False]
    bad = [0]
    got = [0]

    def producer():
        for i in range(ITEMS):
            if gap:
                time.sleep(gap)
            with cv:
                box.append(i)
                cv.notify_all()      # 대기자 전원을 깨운다. 일감은 하나뿐이다
        with cv:
            done[0] = True
            cv.notify_all()

    def consumer():
        while True:
            with cv:
                if recheck:
                    while not box and not done[0]:
                        cv.wait()    # ✅ 깨어날 때마다 조건을 다시 본다
                else:
                    if not box and not done[0]:
                        cv.wait()    # ❌ 깨어났으면 조건이 참이라고 믿는다
                if done[0] and not box:
                    return
                try:
                    box.pop()
                    got[0] += 1
                except IndexError:
                    bad[0] += 1      # 빈 큐에서 꺼냈다 — 통지를 도둑맞았다

    ts = [threading.Thread(target=consumer) for _ in range(CONSUMERS)]
    p = threading.Thread(target=producer)
    for t in ts:
        t.start()
    p.start()
    p.join()
    for t in ts:
        t.join()
    return bad[0], got[0]


def survey(recheck, gap, label):
    rounds_bad = 0
    total_bad = 0
    for _ in range(ROUNDS):
        bad, got = one_round(recheck, gap)
        assert got == ITEMS, (got, ITEMS)
        total_bad += bad
        if bad:
            rounds_bad += 1
    print(f"{label:34s} {rounds_bad:3d}/{ROUNDS} 회 실패   빈 큐 pop 총 {total_bad}회")


if __name__ == "__main__":
    print(f"소비자 {CONSUMERS} / 코어 4 / 항목 {ITEMS}개 / {ROUNDS}회 반복")
    survey(False, 0.0002, "if   판, 생산 간격 0.2ms")
    survey(True, 0.0002, "while 판, 생산 간격 0.2ms")
    survey(False, 0, "if   판, 생산 간격 0 (큐가 안 빈다)")
